metricbeat/helper/elastic/elastic.go

```diff
@@ -158,8 +158,31 @@ func NewModule(base *mb.BaseModule, xpackEnabledMetricsets []string, logger *log
 		return nil, errors.Wrapf(err, "could not unpack configuration for module %v", moduleName)
 	}
 
-	// These metricsets are exactly the ones required if xpack.enabled == true
-	raw["metricsets"] = xpackEnabledMetricsets
+	// Ensure all required metricsets are enabled when xpack.enabled == true
+	cfgdMetricsets, err := raw.GetValue("metricsets")
+	if err != nil || cfgdMetricsets == nil || len(cfgdMetricsets.([]interface{})) == 0 {
+		raw["metricsets"] = xpackEnabledMetricsets
+	} else {
+		// Allow some optional metricsets to be enabled when xpack.enabled == true
+		cfgdMetricsetsSlice := cfgdMetricsets.([]interface{})
+		metricsets := xpackEnabledMetricsets
+
+		for _, cfgdMs := range cfgdMetricsetsSlice {
+			found := false
+			for _, optMs := range optionalXpackMetricsets {
+				if cfgdMs == optMs {
+					found = true
+					break
+				}
+			}
+
+			if found {
+				metricsets = append(metricsets, cfgdMs.(string))
+			}
+		}
+
+		raw["metricsets"] = metricsets
+	}
```

Comment:
I agree with @ruflin's comment https://github.com/elastic/beats/pull/34012#discussion_r1071864218 that this seems to complicated. Open to suggestions here, as I am not a Go expert.

The problem I was having is that `raw.GetValue("metricsets")` returns an `interface{}`. Ideally I could cast it to a `[]string` so I could use `slices.Contains` instead of the inner for loop on line 172. Seems like generics could help here on the `GetValue` but that's a much more involved change.

Reply:
I reworked this a bit and I'm happier with it now.